### ISSUE TITLE
Detect SELinux and mount volumes appropriately

### ISSUE THREAD

**Issue by bollwyvl** [conda-forge/conda-smithy#637]
On a recent Fedora host with SELinux, I can't access the mounted volumes for the feedstock and recipe when running `run_docker_build`.

Looks like recent dockers have [fixed this](http://www.projectatomic.io/blog/2015/06/using-volumes-with-docker-can-cause-problems-with-selinux/).

Because SELinux is a Good Thing, I don't feel like we should punish conscientious developers for running with it.

Haven't gone down the whole rabbit hole, but I think the change would be something like this:
```bash
...
VOLUME_OPTIONS="rw"
if hash getenforce 2> /dev/null && getenforce | grep 'Enforcing' > /dev/null
then
  VOLUME_OPTIONS="${VOLUME_OPTIONS},Z"
fi

cat << EOF | {{ docker.executable }} run -i \
                        -v "${RECIPE_ROOT}":/home/conda/recipe_root:$VOLUME_OPTIONS \
                        -v "${FEEDSTOCK_ROOT}":/home/conda/feedstock_root:$VOLUME_OPTIONS \
                        {{ matrix_env_vars(matrix) }}
                        -a stdin -a stdout -a stderr \
                        {{ docker.image }} \
                        {{ docker.command }} || exit 1
...
```

Barring that, just letting me do `VOLUME_OPTIONS=Z ./scripts/run_docker_build.sh` would do the trick. Happy to PR, with whichever makes the most sense.

**Comment by bollwyvl** [comment]
Indeed, why even mount anything other than `build_artefacts` writeable?

**Comment by jakirkham** [comment]
First I'd like to better understand what you are proposing before we make a change.

Could you help me understand why you want to access the feedstock while the build is running? Not saying there isn't a reason. Just trying to understand your use case is all. :)

Well we actually don't mount `build_artifacts` (recently renamed) as it doesn't necessarily exist. We do mount the feedstock as writable, which is where `build_artifacts` is created. Though we could create `build_artifacts` first and then mount it as writeable. The recipe directory probably can be read only.

Another relevant question is why do we need to mount the recipe and the feedstock directory separately? I don't have a good answer there and we can probably just mount one of them.

What is the `if` statement doing above? Do we need it?

**Comment by bollwyvl** [comment]
My use case is, "locally test the docker build script on an
selinux-enforcing host system". Once inside the container, if the volumes
are not mounted with :z or :Z the recipe can't be found.

Manually adding ':z' to the end of the -v makes it work without complaint.

So looking at what's already there, we're mounting the feedstock and the
recipe (inside the feedstock) as two volumes... I just added the options,
which defaults to 'rw'...

So I'm not doing anything fancy.

Many host systems won't have selinux (and therefore getenforcing), much as
they won't have docker-machine, hence the hash call. Then we grep whether
getenforcing returns the magic string. If all that is true, it augments the
'rw' with ',Z'.

No opinion as to why the current setup is like it is, but yeah: having a
build_artifacts:rw and recipe:ro seems like it would answer the mail.

**Comment by bollwyvl** [comment]
Sorry, feedstock:ro

**Comment by jakirkham** [comment]
Thanks for the follow-up info.

> Once inside the container, if the volumes are not mounted with :z or :Z the recipe can't be found.

I think this is what I was missing. Yes, that is bad. Agree that we should fix it.

> Many host systems won't have selinux (and therefore getenforcing), much as they won't have docker-machine, hence the hash call. Then we grep whether getenforcing returns the magic string. If all that is true, it augments the 'rw' with ',Z'.

Sorry. Should clarify my question further. Are there downsides to adding `,Z` broadly? If so, what are they? Also do we still need `,Z` if we mount with `ro` (this question may make more sense after reading the comment at the end)?

> No opinion as to why the current setup is like it is, but yeah: having a build_artifacts:rw and recipe:ro seems like it would answer the mail.

I think this should be ok. The only downside is one can't edit the recipe in the container (helpful when tightening the feedback loop). Though one could just run their own docker command in that case.

Should add we still need to mount the feedstock no matter what. We need access to the repo for handling merging of PRs that CircleCI does not do and access to some other content. Sorry if that was unclear. We could drop mounting the recipe as well I think (since it is in the feedstock).

Would we be able to mount `build_artifacts` independently as `rw` and the feedstock as `ro` (even though it contains `build_artifacts`)? We could write to `build_artifacts` in the home directory instead of the feedstock directly. The mount structure would be similar to what is going on with the feedstock and recipe now. I'm guessing loosing access to `build_artifacts` on SE Linux during a build is tolerable (if not even desirable).

**Comment by bollwyvl** [comment]
I've been experimenting with a local version vs the gawk feedstock that
prompted this. I also have a Debian box I'll try against.

Indeed, build_art.facts has to exist first, or it will be owned by root.
Mounting the feedstock as read only seems to work fine, if desirable. The
/home/conda/tmpfile issue showed up again, but it could be some docker
cache nonsense.

As to the broader issues with z/Z:
https://github.com/moby/moby/issues/30934

Tldr: So yeah, don't mount ~:/xyz:z. Host Volumes are a leaky abstraction.

Again, I'm fine with just giving a local user such as myself an unloaded
footgun, and not doing magic checking.

**Comment by bollwyvl** [comment]
Here is some concrete changes to the template to demonstrate what I've gotten working:

https://github.com/conda-forge/conda-smithy/pull/640

**Comment by bollwyvl** [comment]
Oh, and my `anvil` was just out-of-date.

**Comment by jakirkham** [comment]
Yeah, I've been doing a lot of hacking on the image of late. ;) None today though.

Does the new image help?

**Comment by jakirkham** [comment]
> The /home/conda/tmpfile issue showed up again, but it could be some docker cache nonsense.

I'm guessing the latest image should fix issues related to this.

**Comment by jakirkham** [comment]
> As to the broader issues with z/Z:
https://github.com/moby/moby/issues/30934
> 
> Tldr: So yeah, don't mount ~:/xyz:z. Host Volumes are a leaky abstraction.

Yikes!

> Again, I'm fine with just giving a local user such as myself an unloaded footgun, and not doing magic checking.

Yeah, I think an environment variable makes sense. This feels like a power user feature to me. Though a power user feature we definitely need.

**Comment by isuruf** [comment]
Any updates on this?

**Comment by bollwyvl** [comment]
I've just been dealing with it locally... Happy to resolve any outstanding
issues...

**Comment by jakirkham** [comment]
I'm a little tempted to table this one until after 3.0.0. There are enough moving parts with `conda-build` 3. Also we are very close to completing that milestone. What do you think @isuruf?

**Comment by bollwyvl** [comment]
No worries from my side: again, I've got it mostly sorted locally, and anybody else that googles `conda-forge` and `selinux` will find this issue... as long as we don't close it. Hooray milestones!

**Comment by jakirkham** [comment]
Thanks @bollwyvl. I’ll leave it unscheduled for now, but would ideally like us to fix a few things and get it in 3.1.0. That all depends on how smoothly 3.0.0 goes. :four_leaf_clover:

**Comment by jakirkham** [comment]
This was ultimately resolved in PR ( https://github.com/conda-forge/conda-smithy/pull/888 ).
